grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomEventLoopGroup.java

```diff
@@ -0,0 +1,182 @@
+package io.grpc.netty.shaded.io.grpc.netty;
+
+import io.grpc.internal.SharedResourcePool;
+import io.grpc.netty.shaded.io.netty.channel.Channel;
+import io.grpc.netty.shaded.io.netty.channel.ChannelFuture;
+import io.grpc.netty.shaded.io.netty.channel.ChannelPromise;
+import io.grpc.netty.shaded.io.netty.channel.EventLoop;
+import io.grpc.netty.shaded.io.netty.channel.EventLoopGroup;
+import io.grpc.netty.shaded.io.netty.util.concurrent.EventExecutor;
+import io.grpc.netty.shaded.io.netty.util.concurrent.Future;
+import io.grpc.netty.shaded.io.netty.util.concurrent.ScheduledFuture;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.jetbrains.annotations.NotNull;
+
+public class CustomEventLoopGroup implements EventLoopGroup {
```

Comment:
Do those event loop groups need to be wrapped? Or could we do something like:

```diff
diff --git a/grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java b/grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java
index 7321d3a6f..001d772a8 100644
--- a/grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java
+++ b/grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java
@@ -1,7 +1,9 @@
 package io.grpc.netty.shaded.io.grpc.netty;
 
+import io.grpc.internal.SharedResourcePool;
 import io.grpc.netty.shaded.io.netty.channel.Channel;
 import io.grpc.netty.shaded.io.netty.channel.ChannelFactory;
+import io.grpc.netty.shaded.io.netty.channel.EventLoopGroup;
 import io.grpc.netty.shaded.io.netty.channel.ServerChannel;
 import io.grpc.netty.shaded.io.netty.channel.group.ChannelGroup;
 import io.grpc.netty.shaded.io.netty.channel.group.DefaultChannelGroup;
@@ -23,6 +25,19 @@ public class CustomChannelFactory implements ChannelFactory<ServerChannel> {
   private final ChannelFactory<? extends ServerChannel> channelFactory;
   private final List<Channel> channels = new ArrayList<>();
 
+
+  public static EventLoopGroup worker() {
+    SharedResourcePool<EventLoopGroup> group =
+        SharedResourcePool.forResource(Utils.DEFAULT_WORKER_EVENT_LOOP_GROUP);
+    return group.getObject();
+  }
+
+  public static EventLoopGroup boss() {
+    SharedResourcePool<EventLoopGroup> group =
+        SharedResourcePool.forResource(Utils.DEFAULT_BOSS_EVENT_LOOP_GROUP);
+    return group.getObject();
+  }
+
   public CustomChannelFactory() {
     this.channelFactory = Utils.DEFAULT_SERVER_CHANNEL_FACTORY;
   }
diff --git a/grpc/src/main/java/io/stargate/grpc/impl/GrpcImpl.java b/grpc/src/main/java/io/stargate/grpc/impl/GrpcImpl.java
index c257b29ca..ad9b53eb7 100644
--- a/grpc/src/main/java/io/stargate/grpc/impl/GrpcImpl.java
+++ b/grpc/src/main/java/io/stargate/grpc/impl/GrpcImpl.java
@@ -18,7 +18,6 @@ package io.stargate.grpc.impl;
 import io.grpc.Server;
 import io.grpc.internal.GrpcUtil;
 import io.grpc.netty.shaded.io.grpc.netty.CustomChannelFactory;
-import io.grpc.netty.shaded.io.grpc.netty.CustomEventLoopGroup;
 import io.grpc.netty.shaded.io.grpc.netty.NettyServerBuilder;
 import io.micrometer.core.instrument.binder.grpc.MetricCollectingServerInterceptor;
 import io.stargate.auth.AuthenticationService;
@@ -81,8 +80,8 @@ public class GrpcImpl {
             // Otherwise, it will throw: All of BossEventLoopGroup, WorkerEventLoopGroup and
             // ChannelType should be provided or neither should be
             .channelFactory(customChannelFactory)
-            .workerEventLoopGroup(CustomEventLoopGroup.worker())
-            .bossEventLoopGroup(CustomEventLoopGroup.boss())
+            .workerEventLoopGroup(CustomChannelFactory.worker())
+            .bossEventLoopGroup(CustomChannelFactory.boss())
             .build();
 
     persistence.registerEventListener(new EventNotifier(customChannelFactory));
diff --git a/grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java b/grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java
index 7321d3a6f..001d772a8 100644
--- a/grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java
+++ b/grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java
@@ -1,7 +1,9 @@
 package io.grpc.netty.shaded.io.grpc.netty;
 
+import io.grpc.internal.SharedResourcePool;
 import io.grpc.netty.shaded.io.netty.channel.Channel;
 import io.grpc.netty.shaded.io.netty.channel.ChannelFactory;
+import io.grpc.netty.shaded.io.netty.channel.EventLoopGroup;
 import io.grpc.netty.shaded.io.netty.channel.ServerChannel;
 import io.grpc.netty.shaded.io.netty.channel.group.ChannelGroup;
 import io.grpc.netty.shaded.io.netty.channel.group.DefaultChannelGroup;
@@ -23,6 +25,19 @@ public class CustomChannelFactory implements ChannelFactory<ServerChannel> {
   private final ChannelFactory<? extends ServerChannel> channelFactory;
   private final List<Channel> channels = new ArrayList<>();
 
+
+  public static EventLoopGroup worker() {
+    SharedResourcePool<EventLoopGroup> group =
+        SharedResourcePool.forResource(Utils.DEFAULT_WORKER_EVENT_LOOP_GROUP);
+    return group.getObject();
+  }
+
+  public static EventLoopGroup boss() {
+    SharedResourcePool<EventLoopGroup> group =
+        SharedResourcePool.forResource(Utils.DEFAULT_BOSS_EVENT_LOOP_GROUP);
+    return group.getObject();
+  }
+
   public CustomChannelFactory() {
     this.channelFactory = Utils.DEFAULT_SERVER_CHANNEL_FACTORY;
   }
diff --git a/grpc/src/main/java/io/stargate/grpc/impl/GrpcImpl.java b/grpc/src/main/java/io/stargate/grpc/impl/GrpcImpl.java
index c257b29ca..ad9b53eb7 100644
--- a/grpc/src/main/java/io/stargate/grpc/impl/GrpcImpl.java
+++ b/grpc/src/main/java/io/stargate/grpc/impl/GrpcImpl.java
@@ -18,7 +18,6 @@ package io.stargate.grpc.impl;
 import io.grpc.Server;
 import io.grpc.internal.GrpcUtil;
 import io.grpc.netty.shaded.io.grpc.netty.CustomChannelFactory;
-import io.grpc.netty.shaded.io.grpc.netty.CustomEventLoopGroup;
 import io.grpc.netty.shaded.io.grpc.netty.NettyServerBuilder;
 import io.micrometer.core.instrument.binder.grpc.MetricCollectingServerInterceptor;
 import io.stargate.auth.AuthenticationService;
@@ -81,8 +80,8 @@ public class GrpcImpl {
             // Otherwise, it will throw: All of BossEventLoopGroup, WorkerEventLoopGroup and
             // ChannelType should be provided or neither should be
             .channelFactory(customChannelFactory)
-            .workerEventLoopGroup(CustomEventLoopGroup.worker())
-            .bossEventLoopGroup(CustomEventLoopGroup.boss())
+            .workerEventLoopGroup(CustomChannelFactory.worker())
+            .bossEventLoopGroup(CustomChannelFactory.boss())
             .build();
 
     persistence.registerEventListener(new EventNotifier(customChannelFactory));
```

Then this class could be removed.

Reply:
The class needs to be different than: `Utils.DEFAULT_BOSS_EVENT_LOOP_GROUP` so wrapping it using:
`SharedResourcePool.forResource(Utils.DEFAULT_WORKER_EVENT_LOOP_GROUP);` works. Thx!